aws/resource_aws_ec2_traffic_mirror_session.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	if d.HasChange("session_number") {
+		_, n := d.GetChange("session_number")
+		input.SessionNumber = aws.Int64(int64(n.(int)))
+	}
+
+	if d.HasChange("traffic_mirror_filter_id") {
+		_, n := d.GetChange("traffic_mirror_filter_id")
+		input.TrafficMirrorFilterId = aws.String(n.(string))
+	}
+
+	if d.HasChange("traffic_mirror_target_id") {
+		_, n := d.GetChange("traffic_mirror_target_id")
+		input.TrafficMirrorTargetId = aws.String(n.(string))
+	}
+
+	var removeFields []*string
+	if d.HasChange("description") {
+		_, n := d.GetChange("description")
+		if "" != n {
+			input.Description = aws.String(n.(string))
+		} else {
+			removeFields = append(removeFields, aws.String("description"))
+		}
+	}
+
+	if d.HasChange("packet_length") {
+		_, n := d.GetChange("packet_length")
+		if nil != n && n.(int) > 0 {
+			input.PacketLength = aws.Int64(int64(n.(int)))
+		} else {
+			removeFields = append(removeFields, aws.String("packet-length"))
+		}
+	}
+	log.Printf("[DEBUG] removeFields %v", removeFields)
+
+	if d.HasChange("virtual_network_id") {
+		_, n := d.GetChange("virtual_network_id")
+		log.Printf("[DEBUG] VNI has change %v", n)
+		if nil != n && n.(int) > 0 {
+			input.VirtualNetworkId = aws.Int64(int64(n.(int)))
+		} else {
+			removeFields = append(removeFields, aws.String("virtual-network-id"))
+		}
+	}
+
+	log.Printf("[DEBUG] removeFields %v", removeFields)
+	if len(removeFields) > 0 {
+		input.SetRemoveFields(removeFields)
+	}
+
+	_, err := conn.ModifyTrafficMirrorSession(input)
+	if nil != err {
+		return fmt.Errorf("Error updating traffic mirror session %v", err)
+	}
+
+	return resourceAwsEc2TrafficMirrorSessionRead(d, meta)
+}
+
+func resourceAwsEc2TrafficMirrorSessionRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	sessionId := d.Id()
+	input := &ec2.DescribeTrafficMirrorSessionsInput{
+		TrafficMirrorSessionIds: []*string{
+			&sessionId,
+		},
+	}
+
+	out, err := conn.DescribeTrafficMirrorSessions(input)
+	if nil != err {
+		d.SetId("")
+		return fmt.Errorf("Error describing traffic mirror session %v", sessionId)
+	}
+
+	if 0 == len(out.TrafficMirrorSessions) {
+		d.SetId("")
+		return fmt.Errorf("Unable to find traffic mirrir session %v", sessionId)
+	}
+
+	session := out.TrafficMirrorSessions[0]
+	d.Set("network_interface_id", session.NetworkInterfaceId)
+	d.Set("session_number", session.SessionNumber)
+	d.Set("traffic_mirror_filter_id", session.TrafficMirrorFilterId)
+	d.Set("traffic_mirror_target_id", session.TrafficMirrorTargetId)
+	d.Set("description", session.Description)
+	d.Set("packet_length", session.PacketLength)
+
+	// AWS returns a value for VNI even if nothing is set in config. To avoid further plan
+	// showing a change, using a workaround - set this in state only if it was set in config
+	_, vniSet := d.GetOk("virtual_network_id")
+	if vniSet && nil != session.VirtualNetworkId {
+		d.Set("virtual_network_id", *session.VirtualNetworkId)
+	}
```

Comment:
We should always prefer to call `d.Set()` in Read functions -- `d.Get()` and `d.GetOk()` logic does not work during import. If this configuration is optional and is defaulted by the API in that case, then setting `Computed: true` on the schema attribute is more appropriate.

```go
			"virtual_network_id": {
				Type:         schema.TypeInt,
				Optional:     true,
				Computed:     true,
				ValidateFunc: validation.IntBetween(1, 16777216),
			},
```

And the logic here reduced to just:

```go
d.Set("virtual_network_id", session.VirtualNetworkId)
```